File: Backend/src/main/java/com/maskan/api/service/BookingService.java
package com.maskan.api.service;

import com.maskan.api.dto.BookingRequest;
import com.maskan.api.dto.BookingResponse;
import com.maskan.api.dto.BookingStatusUpdateRequest;
import com.maskan.api.dto.CheckInVerificationResponse;
import com.maskan.api.dto.VerifyCheckInRequest;
import com.maskan.api.dto.BookedDateRangeResponse;
import com.maskan.api.dto.UnavailableDateRangeResponse;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface BookingService {
    BookingResponse createBooking(BookingRequest request, String email);
    BookingResponse updateStatus(String bookingId, BookingStatusUpdateRequest request, String email);
    void cancelBooking(String bookingId, String email);
    CheckInVerificationResponse verifyCheckIn(String bookingId, VerifyCheckInRequest request, String email);
    Page<BookingResponse> getMyBookings(String email, Pageable pageable);
    Page<BookingResponse> getOwnerBookings(String email, Pageable pageable);
    Page<BookingResponse> getAllBookings(Pageable pageable);
    List<UnavailableDateRangeResponse> getUnavailableDateRangesForListing(String listingId);
    List<BookedDateRangeResponse> getBookedDateRangesForProperty(String propertyId);
}

File: Frontend/src/services/bookingService.ts
import { apiClient } from '../api/apiClient'
import { ENDPOINTS } from '../api/endpoints'
import { buildQueryString } from '../api/query'
import type {
  BookingRequest,
  BookingResponse,
  BookingStatusUpdateRequest,
  CheckInVerificationResponse,
  PageResponse,
  PaymentCheckoutResponse,
  UnavailableDateRange,
  BookedDateRange,
} from '../utils/contracts'

export const bookingService = {
  async create(payload: BookingRequest): Promise<BookingResponse> {
    const { data } = await apiClient.post<BookingResponse>(ENDPOINTS.bookings.create, payload)
    return data
  },

  async getMine(query: { page?: number; size?: number; sort?: string } = {}): Promise<PageResponse<BookingResponse>> {
    const queryString = buildQueryString(query as Record<string, unknown>)
    const { data } = await apiClient.get<PageResponse<BookingResponse>>(`${ENDPOINTS.bookings.listMine}${queryString}`)
    return data
  },

  async getOwnerBookings(query: { page?: number; size?: number; sort?: string } = {}): Promise<PageResponse<BookingResponse>> {
    const queryString = buildQueryString(query as Record<string, unknown>)
    const { data } = await apiClient.get<PageResponse<BookingResponse>>(`${ENDPOINTS.bookings.listOwner}${queryString}`)
    return data
  },

  async getUnavailableDates(listingId: number | string): Promise<UnavailableDateRange[]> {
    const { data } = await apiClient.get<UnavailableDateRange[]>(ENDPOINTS.bookings.unavailableDates(listingId))
    return data
  },

  async getBookedDates(propertyId: number | string): Promise<BookedDateRange[]> {
    const { data } = await apiClient.get<BookedDateRange[]>(ENDPOINTS.properties.bookedDates(propertyId))
    return data
  },

  async updateStatus(id: number | string, payload: BookingStatusUpdateRequest): Promise<BookingResponse> {
    const { data } = await apiClient.patch<BookingResponse>(ENDPOINTS.bookings.updateStatus(id), payload)
    return data
  },

  async payEscrow(bookingId: number | string): Promise<PaymentCheckoutResponse> {
    const { data } = await apiClient.put<PaymentCheckoutResponse>(ENDPOINTS.bookings.pay(bookingId))
    return data
  },

  async checkoutPayment(bookingId: number | string): Promise<PaymentCheckoutResponse> {
    return this.payEscrow(bookingId)
  },

  async verifyCheckIn(bookingId: number | string, secretCode: string): Promise<CheckInVerificationResponse> {
    const { data } = await apiClient.post<CheckInVerificationResponse>(ENDPOINTS.bookings.verifyCheckIn(bookingId), { secretCode })
    return data
  },
}
